### ISSUE TITLE
hidden vector while training the language model

### ISSUE THREAD

**Issue by bobshih** [flairNLP/flair#1062]
Hello,

recently, I have studied this repository.
When tracing training process of language model, I found that hidden vector is passed all time, but the hidden vector is changing and not the initial one.
[here](https://github.com/zalandoresearch/flair/blob/005ae9067e8052c0fbd94c6a3d18068477e19317/flair/trainers/language_model_trainer.py#L377) produce the new hidden vector which is different from the initial one.
[here](https://github.com/zalandoresearch/flair/blob/005ae9067e8052c0fbd94c6a3d18068477e19317/flair/trainers/language_model_trainer.py#L397) copies and detach this hidden vector, which is changed in this batch.

My question is that if the hidden vector is always changing, how to ask the LSTM to produce the same output for the same input?
If I miss something which can help me realize the process, please tell me.
I am appreciate your answers.

Finally, sorry for my poor English. I am not a native English speaker.

**Comment by alanakbik** [comment]
Hello @bobshih the training technique is called "truncated backpropagation through time". This helps address the problem of long sequences. Essentially, instead of putting a very long sequence of text (like a book or a corpus) as one long sequence through an LSTM, you split the sequence up into blocks and always pass the last hidden state of the last block to initialize the next block. Then at least in the forward pass the information can travel through the whole sequence. The detaching of gradients means that the gradient can only go back within the same block, which means that the RNN only has to be unrolled for a limited number of steps, speeding up computation.

**Comment by bobshih** [comment]
Hi, @alanakbik 
Thanks for your answer.
After studying the truncated BPTT, I have realized that this technique is used to solve the problem that  long sequences are put in LSTM.
And there is another question raising.
Language model predicts the P(Xt|Xt-1...Xt-n)
Training the language model with long sequences such as a book or a paragraph is a good way?
In my understanding, training the language model uses the sentences instead of books or paragraphs.
I have briefly studied paper introducing the Flair, and I am wondering the reason why using the truncated BPTT is that Flair is trained with characters which will make a sentence  become a much longer character sequence.
Please forgive me if i ask something stupid.
